### PR TITLE
[Fix] Fix FA2 kernels ut

### PR DESCRIPTION
**Pull request by YangKai0616**
# What does this PR do?

This PR fixes the UTs where FA2 falls back to kernels loading.
For example `pytest -rA tests/models/sam3/test_modeling_sam3.py::Sam3ModelTest::test_flash_attn_2_can_dispatch_composite_models`.

### CODE REVIEW

**Comment by YangKai0616** [comment]
Hi @vasqu , please help review, thanks!

File: tests/test_modeling_common.py [hunk 1]
```diff
@@ -3248,12 +3248,30 @@ def flash_attn_can_dispatch_composite_models(self, attn_implementation: str):
...
+            # Allow kernels fallbacks for flash attention tests.
+            requested = attn_impl
+            base = requested.removeprefix("paged|")
+            prefix = "paged|" if requested.startswith("paged|") else ""
+
+            expected = {requested}
+            kernel_map = {
+                "flash_attention_2": "kernels-community/flash-attn2",
+                "flash_attention_3": "kernels-community/vllm-flash-attn3",
+            }
```
Comment (vasqu) [inline tests/test_modeling_common.py]: Could be something static at some point, currently we do these manually, e.g. modeling utils has a (private) constant which we access. Not necessarily here, just a thought

File: tests/test_modeling_common.py [hunk 2]
```diff
@@ -3248,12 +3248,30 @@ def flash_attn_can_dispatch_composite_models(self, attn_implementation: str):
             self.skipTest(f"bfloat16 not supported on {torch_device} (on the specific device currently used)")
 
         dtype = torch.bfloat16
+
+        def _expected_attn_impls(attn_impl: str) -> set[str]:
+            # Allow kernels fallbacks for flash attention tests.
+            requested = attn_impl
```
Comment (vasqu) [inline tests/test_modeling_common.py]: Let's be explicit and go for `attention_implementation`

File: tests/test_modeling_common.py [hunk 3]
```diff
@@ -3248,12 +3248,30 @@ def flash_attn_can_dispatch_composite_models(self, attn_implementation: str):
...
+                expected.add(f"{prefix}{kernel_map[base]}")
+            return expected
+
         for model_class in self.all_model_classes:
             config, inputs_dict = self.model_tester.prepare_config_and_inputs_for_common()
             model = model_class(config)
             if not self._is_composite:
                 self.skipTest("This model is not a composite model!")
 
+            expected_attn_impls = _expected_attn_impls(attn_implementation)
```
Comment (vasqu) [inline tests/test_modeling_common.py]: Can be moved before the for loop, no?

**Review by vasqu — COMMENTED** [review]
Just some smaller comments, dw about the CI too much we work on getting it green again (should be unrelated to your PR)

File: tests/test_modeling_common.py (hunk 1 above)
Comment (YangKai0616) [reply tests/test_modeling_common.py]: Got it. I've refactored the relevant code. Could you please review it and point out any shortcomings or areas for improvement?

File: tests/test_modeling_common.py (hunk 2 above)
Comment (YangKai0616) [reply tests/test_modeling_common.py]: Done.

File: tests/test_modeling_common.py (hunk 3 above)
Comment (YangKai0616) [reply tests/test_modeling_common.py]: Done!

File: src/transformers/modeling_flash_attention_utils.py [hunk 4]
```diff
@@ -112,9 +112,11 @@ def _lazy_imports(implementation: Optional[str], attention_wrapper: Optional[Cal
             flash_attn_func = getattr(kernel, "flash_attn_func", None)
             flash_attn_varlen_func = getattr(kernel, "flash_attn_varlen_func", None)
             if flash_attn_varlen_func is None:
+                from .modeling_utils import _FLASH_ATTN_KERNEL_FALLBACK
+
                 raise ValueError(
                     f"Could not find the currently requested flash attention implementation at `{implementation}`."
-                    "Make sure that you request a valid kernel from the hub, e.g. `kernels-community/flash-attn2`."
+                    f"Make sure that you request a valid kernel from the hub, e.g. `{_FLASH_ATTN_KERNEL_FALLBACK['flash_attention_2']}`."
```
Comment (vasqu) [inline src/transformers/modeling_flash_attention_utils.py]: I think we can keep it as is, this is only an example so no need it import just for this

File: src/transformers/modeling_utils.py [hunk 5]
```diff
@@ -156,6 +156,12 @@
 _is_quantized = False
 _is_ds_init_called = False
 
+# Mapping from flash attention implementations to their kernel fallback repositories
+_FLASH_ATTN_KERNEL_FALLBACK = {
+    "flash_attention_2": "kernels-community/flash-attn2",
+    "flash_attention_3": "kernels-community/vllm-flash-attn3",
+}
```
Comment (vasqu) [inline src/transformers/modeling_utils.py]:
```suggestion
FLASH_ATTN_KERNEL_FALLBACK = {
    "flash_attention_2": "kernels-community/flash-attn2",
    "flash_attention_3": "kernels-community/vllm-flash-attn3",
}
```
I think it's fair to make it non-private

File: tests/test_modeling_common.py [hunk 6]
```diff
@@ -2749,12 +2749,6 @@ def flash_attn_inference_equivalence(
             if getattr(config, "sliding_window", None):
                 config.sliding_window = 2
 
-                if torch_device == "xpu" and (
-                    attn_implementation == "kernels-community/flash-attn2"
-                    or attn_implementation == "flash_attention_2"
-                ):
-                    self.skipTest("XPU does not support sliding window attention with Flash-Attention-2 currently.")
```
Comment (vasqu) [inline tests/test_modeling_common.py]: Ah this works now I assume?

File: tests/utils/test_modeling_utils.py [hunk 7]
```diff
@@ -129,7 +129,11 @@
         _prepare_4d_attention_mask,
         _prepare_4d_causal_attention_mask,
     )
-    from transformers.modeling_utils import _find_disjoint, _find_identical
+    from transformers.modeling_utils import (
+        _FLASH_ATTN_KERNEL_FALLBACK,
+        _find_disjoint,
+        _find_identical,
```
Comment (vasqu) [inline tests/utils/test_modeling_utils.py:135]: Unnecessary imports? (`_find_disjoint`, `_find_identical`)

**Review by vasqu — APPROVED** [review]
LGTM, just a few nits left

File: src/transformers/modeling_flash_attention_utils.py (hunk 4 above)
Comment (YangKai0616) [reply src/transformers/modeling_flash_attention_utils.py]: Done.

File: src/transformers/modeling_utils.py (hunk 5 above)
Comment (YangKai0616) [reply src/transformers/modeling_utils.py]: Done.

File: tests/test_modeling_common.py (hunk 6 above)
Comment (YangKai0616) [reply tests/test_modeling_common.py]: Yes, the FA2 of XPU now supports local mask. I removed it in the recent PR https://github.com/huggingface/transformers/pull/42798. Sry I forgot to mention it here.

File: tests/utils/test_modeling_utils.py (hunk 7 above)
Comment (YangKai0616) [reply tests/utils/test_modeling_utils.py:135]: It seems that [here](https://github.com/huggingface/transformers/blob/40dc11cd3eb4126652aa41ef8272525affd4a636/tests/utils/test_modeling_utils.py#L3050C1-L3050C39) is using these two methods.

File: tests/test_modeling_common.py (hunk 6 above)
Comment (vasqu) [reply tests/test_modeling_common.py]: No worries

File: tests/utils/test_modeling_utils.py (hunk 7 above)
Comment (vasqu) [reply tests/utils/test_modeling_utils.py:135]: You're right, I didnt notice it was just reformatting the imports here :facepalm:
